extensions/ql-vscode/src/remote-queries/variant-analysis-monitor.ts

```diff
@@ -0,0 +1,81 @@
+import * as vscode from 'vscode';
+import { Credentials } from '../authentication';
+import { Logger } from '../logging';
+import * as ghApiClient from './gh-api/gh-api-client';
+
+import { VariantAnalysis, VariantAnalysisStatus } from './shared/variant-analysis';
+import {
+  VariantAnalysis as VariantAnalysisApiResponse
+} from './gh-api/variant-analysis';
+import { VariantAnalysisMonitorResult } from './shared/variant-analysis-monitor-result';
+import { processFailureReason } from './variant-analysis-processor';
+
+export class VariantAnalysisMonitor {
+  // With a sleep of 5 seconds, the maximum number of attempts takes
+  // us to just over 2 days worth of monitoring.
+  public static maxAttemptCount = 17280;
+  public static sleepTime = 5000;
+
+  constructor(
+    private readonly extensionContext: vscode.ExtensionContext,
+    private readonly logger: Logger
+  ) {
+  }
+
+  public async monitorVariantAnalysis(
+    variantAnalysis: VariantAnalysis,
+    cancellationToken: vscode.CancellationToken
+  ): Promise<VariantAnalysisMonitorResult> {
+
+    const credentials = await Credentials.initialize(this.extensionContext);
+    if (!credentials) {
+      throw Error('Error authenticating with GitHub');
+    }
+
+    let variantAnalysisSummary: VariantAnalysisApiResponse;
+    let attemptCount = 0;
+    const scannedReposDownloaded: number[] = [];
+
+    while (attemptCount <= VariantAnalysisMonitor.maxAttemptCount) {
+      await this.sleep(VariantAnalysisMonitor.sleepTime);
+
+      if (cancellationToken && cancellationToken.isCancellationRequested) {
+        return { status: 'Cancelled', error: 'Variant Analysis was canceled.' };
+      }
+
+      variantAnalysisSummary = await ghApiClient.getVariantAnalysis(
```

Comment:
Does it make sense to move the variable declaration to inside the `while` loop? It doesn't seem to be used outside of it, so making it a `const` would make this somewhat more robust.

Reply:
Your observation is spot on. At the moment there's not much point to declaring it outside the while loop. 

However, I've left it like this since I intend to use it in a future PR in order to check on the summary's finished stated (e.g. did it time out or fail?). You can see a sketch of this in one of my other branches: https://github.com/github/vscode-codeql/compare/elenatanasoiu/monitor-variant-analysis-continued#diff-65d29f358913c786586c8fdf81a804c51bc1853333b8b02ccc6c950a4f0ee757R88-R98.

Reply:
Just for the sake of clarity, I had a chat with Robert about what this monitor is meant to return and based on what he described in [the comment](https://github.com/github/vscode-codeql/pull/1545#discussion_r982591809) below, we decided not to return anything from the monitor and instead rely on emitting events to the UI for updates on the state (we'll refactor this in a future PR). 

We'd also like to trigger separate commands for actually downloading the results rather than performing it in the same process as the monitor. The reason is because it would delay the polling and isn't really connected to what the monitor is supposed to do.